manifestservice/manifests/__init__.py

```diff
@@ -0,0 +1,311 @@
+    False otherwise.
+    """
+    privileges = list(project_access_dict.values())
+
+    for auth_set in privileges:
+        if "read" in auth_set and "read-storage" in auth_set:
+            return True
+
+    return False
+
+
+def is_valid_manifest(manifest_json, required_keys):
+    """
+    Returns True if the manifest.json is a list of the form [{'k' : v}, ...], 
+    where each member dictionary contains an object_id key.
+    Otherwise, returns False
+    """
+    for record in manifest_json:
+        record_keys = record.keys()
+        if not set(required_keys).issubset(record_keys):
+            return False
+
+    return True
+
+
+def _generate_unique_manifest_filename(
+    folder_name, manifest_bucket_name, users_existing_manifest_files
+):
+    """
+    Returns a filename of the form manifest-<timestamp>-<optional-increment>.json that is
+    unique among the files in the user's manifest folder.
+    """
+    timestamp = datetime.now().isoformat()
+    existing_filenames = map(lambda x: x["filename"], users_existing_manifest_files)
+    filename = _generate_unique_filename_with_timestamp_and_increment(
+        timestamp, existing_filenames
+    )
+    return filename
+
+
+def _generate_unique_filename_with_timestamp_and_increment(
+    timestamp, users_existing_manifest_files
+):
+    """
+    A helper function for _generate_unique_manifest_filename(), which facilitates unit testing.
+    Adds an increment to the filename if there happens to be another timestamped file with the same name
+    (unlikely, but good to check). 
+    """
+    filename_without_extension = "manifest-" + timestamp.replace(":", "-")
+    extension = ".json"
+
+    filename = filename_without_extension + extension
+    i = 1
+    while filename in users_existing_manifest_files:
+        filename = filename_without_extension + extension
+        if filename in users_existing_manifest_files:
+            filename = filename_without_extension + "-" + str(i) + extension
+        i += 1
+
+    return filename
+
+
+def _list_files_in_bucket(bucket_name, folder):
+    """
+    Lists the files in an s3 bucket. Returns a list of filenames.
+    """
+    session = boto3.Session(
+        region_name="us-east-1",
+        aws_access_key_id=app.config["AWS_ACCESS_KEY_ID"],
+        aws_secret_access_key=app.config["AWS_SECRET_ACCESS_KEY"],
+    )
+    s3 = session.resource("s3")
+
+    rv = []
+    bucket = s3.Bucket(bucket_name)
+
+    try:
+        bucket_objects = bucket.objects.filter(Prefix=folder + "/")
+        for object_summary in bucket_objects:
+            manifest_summary = {
+                "filename": ntpath.basename(object_summary.key),
+                "last_modified": object_summary.last_modified.strftime(
+                    "%Y-%m-%d %H:%M:%S"
+                ),
+            }
+            rv.append(manifest_summary)
+    except Exception as e:
+        print(e)
+        return str(e), False
+
+    return rv, True
```

Comment:
The reason I went for this pattern here was because I don't want to throw this particular AWS exception -- I want to catch it, log it, and give the user a different error message. The error messages are being delivered to a UI component which will display the error to the end user in their browser. I don't want a STAGE user to see our stack trace when they try exporting data files.

Reply:
Also, this function is prefaced with an underscore -- the signature for this function is encapsulated :P